### DISCUSSION TITLE
Import, unreachable after applying the values in `DefinePlugin`, appears in built product anyway

### DISCUSSION THREAD

**Discussion by steveluscher** [webpack/webpack#16980] [Q&A]
# Bug report

**What is the current behavior?**

Observe that when `__BROWSER__` is made `true` through the use of `DefinePlugin`, this code:

```ts
import nodeFetchImpl from "node-fetch";

let fetchImpl;
if (__BROWSER__) {
  fetchImpl = globalThis.fetch;
} else {
  fetchImpl = nodeFetchImpl;
}

export default fetchImpl;
```

…gets transformed to this:

```js
import*as e from"node-fetch";var r={544:r=>{r.exports=e}},t={};function o(e){var a=t[e];if(void 0!==a)return a.exports;var s=t[e]={exports:{}};return r[e](s,s.exports,o),s.exports}var a,s,p={};s=p,Object.defineProperty(s,"X",{value:!0}),o(544),a=globalThis.fetch,s.Z=a;var f=p.X,v=p.Z;export{f as __esModule,v as default};
```

**If the current behavior is a bug, please provide the steps to reproduce.**

Repro: https://github.com/steveluscher/webpack-dce-repro

**What is the expected behavior?**

Since:

1. `node-fetch` has `"sideEffects": false` in its `package.json`, and
2. The `node-fetch` import is unreachable when `__BROWSER__` is `true`

I would expect the compiled output not to contain `import*as e from"node-fetch";` or any of the code associated with `e`.

**Other relevant information:**

```
Binaries:
  Node: 16.20.0 - /usr/local/bin/node
  npm: 8.19.4 - /usr/local/bin/npm
Packages:
  ts-loader: ^9.4.2 => 9.4.2 
  webpack: ^5.79.0 => 5.79.0 
  webpack-cli: ^5.0.1 => 5.0.1 
```

**Comment by alexander-akait** [comment]
Everything works as expected:

> node-fetch has "sideEffects": false in its package.json, and

You have `import * as __WEBPACK_EXTERNAL_MODULE_node_fetch_e2fb16d6__ from "node-fetch";` on the top, because it is external and `sideEffects` will not work in such case

**Comment by alexander-akait** [comment]
Anyway you can do this using:
```ts
const fetchImpl = (async () => {
  let fetchImpl;
  if (__BROWSER__) {
    fetchImpl = globalThis.fetch;
  } else {
    fetchImpl = await import("node-fetch");
  }

  return fetchImpl;
})();

export default fetchImpl;
```

```
externalsType: "import",
externals: ["node-fetch"],
```

**Comment by alexander-akait** [comment]
You need `externalsType` due this https://github.com/webpack/webpack/issues/16181 we will fix it in future

**Comment by steveluscher** [comment]
> Everything works as expected…

Not yet – not for me anyway. The [Webpack docs](https://github.com/webpack/webpack/blob/main/examples/side-effects/README.md) have this to say about `sideEffects`:

> The "sideEffects": false flag in big-module's package.json indicates that the package's modules have no side effects (on evaluation) and only expose exports. This allows tools like webpack to optimize re-exports.

I don't think it's controversial to interpret that flag as an indicator that – absent a use – an import can be reaped. If `node-fetch` was not marked as `"sideEffects": false` and Webpack left this code alone:

```ts
require('node-fetch');
```

…that would be understandable! We can't know that the _module factory_ of `node-fetch` is side-effect free (eg. it _might_ overwrite `globalThis.fetch`) so we _have_ to leave it alone, to be safe.

That's not what's going on here. `node-fetch` is marked as side-effect free, it's not used anywhere in the body of the compiled `index.ts`, but despite all of that, Webpack doesn't reap it.

Why?

**Comment by steveluscher** [comment]
> Anyway you can do this using:
> 
> ```ts
> const fetchImpl = (async () => {
>   let fetchImpl;
>   if (__BROWSER__) {
>     fetchImpl = globalThis.fetch;
>   } else {
>     fetchImpl = await import("node-fetch");
>   }
> 
>   return fetchImpl;
> })();
> 
> export default fetchImpl;
> ```
> 
> ```
> externalsType: "import",
> externals: ["node-fetch"],
> ```

That code exports a Promise for a function, rather than a function, so that will not work. Also, `import("node-fetch")` still appears in the built output, so it doesn't achieve the goal.

**Comment by alexander-akait** [comment]
> node-fetch is marked as side-effect free
> That's not what's going on here. node-fetch is marked as side-effect free, it's not used anywhere in the body of the compiled index.ts, but despite all of that, Webpack doesn't reap it.

It is not applied for externals, because we don't know what is `node-fetch`, it can be a package, it can be a variable in global scope, it can be anything in scope and can be loaded from `window`, `network` (using http/https) and etc, so we can't undestand it is `sideEffects` free, so it will not work and never worked




> That code exports a Promise for a function, rather than a function, so that will not work. Also, import("node-fetch") still appears in the built output, so it doesn't achieve the goal.

Try to run it again, maybe you missing something, all works perfectly locally, it remove `node-fetch`, but yes, it resturns promise, you can await it too when you need or using top level awaits and don't have it at all

**Reply by steveluscher** [reply]
> It is not applied for externals, because we don't know what is `node-fetch`…

I don't understand this. When you write `require('foo')` or `import X from 'foo'`, then `foo` is uncontroversially a module, right? Especially considering it has an entry in `package.json` under `dependencies`. The ESLint unused import linter, for instance, has no trouble identifying this require/import as unused. What is to stop Webpack from doing the same?

**Reply by steveluscher** [reply]
> …you can await it too when you need…

I can't. This is a shim for the `fetch` API. I can't change it from sync to async at will, because downstream consumers won't be compatible unless its sync.

**Comment by alexander-akait** [comment]
For ESM webpack by default move externals at top due https://github.com/webpack/webpack/issues/16181 (yes we want to fix it in future) and for externals we don't check `sideEffects`, that is why you have `node-fetch` at the top and it is not removed, the solution above, you need using async functions and `import(...)`, also you can rewrite it on top level awaits

**Reply by steveluscher** [reply]
You seem to have indicated that this is a known problem, and fixable. I'm disappointed in your decision to have migrated this issue to a GitHub discussion. Now we can't track it as a feature request.

**Reply by alexander-akait** [reply]
@steveluscher I'm afraid it's not entirely possible without rewriting the code to async (currently), because as I said above externals are not checking on `sideEffects`, so we can't just drop it, because it is unsafe, I reopen an another issue - https://github.com/webpack/webpack/issues/15486, because you are not the first who asked about it, so let's track it there, anyway PR welcome, it should be not hard to implement
